### ISSUE TITLE
mitigate GPPCA aborts

### ISSUE THREAD

**Issue by davidsebfischer** [theislab/cellrank#277]
- [x] Additional function parameters, change functionality or change defaults?
- [ ] New estimator in `cr.tl.estimators`?
- [ ] New kernel in `cr.tl.kernels`?
- [ ] New gene trend model in `cr.ul.models`?
- [ ] New plotting function in `cr.pl`?
- [ ] Other?

i m getting this error in some data partitions when I run compute_metastable_states on GPCCA instantiations with fixed n_states:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-99-22fd2b57c6f4> in <module>
     64                 cluster_key="louvain",
     65                 en_cutoff=0.7,
---> 66                 p_thresh=1e-15
     67             )
     68             estim.set_main_states()

~/opt/miniconda3/lib/python3.7/site-packages/cellrank/tools/estimators/_gppca.py in compute_metastable_states(self, n_states, use_min_chi, n_cells, cluster_key, en_cutoff, p_thresh)
    471         except ValueError as e:
    472             if n_states != self._schur_vectors.shape[1]:
--> 473                 raise e
    474             logg.warning(
    475                 f"Unable to perform the optimization using `{self._schur_vectors.shape[1]}` Schur vectors. "

~/opt/miniconda3/lib/python3.7/site-packages/cellrank/tools/estimators/_gppca.py in compute_metastable_states(self, n_states, use_min_chi, n_cells, cluster_key, en_cutoff, p_thresh)
    468         start = logg.info("Computing metastable states")
    469         try:
--> 470             self._gpcca = self._gpcca.optimize(m=n_states)
    471         except ValueError as e:
    472             if n_states != self._schur_vectors.shape[1]:

~/opt/miniconda3/lib/python3.7/site-packages/cellrank/_vendor/msmtools/analysis/dense/gpcca.py in optimize(self, m, return_extra)
   1276 
   1277         # Calculate Schur matrix R and Schur vector matrix X, if not adequately given.
-> 1278         self._do_schur_helper(max(m_list))
   1279 
   1280         # Initialize lists to collect results.

~/opt/miniconda3/lib/python3.7/site-packages/cellrank/_vendor/msmtools/analysis/dense/gpcca.py in _do_schur_helper(self, m)
   1075                     else:
   1076                         if _check_conj_split(self.eigenvalues[:m]):
-> 1077                             raise ValueError(f'Clustering into {m} clusters will split conjugate eigenvalues. '
   1078                                              f'Request one cluster more or less. ')
   1079                         print('INFO: Using pre-computed schur decomposition')

ValueError: Clustering into 3 clusters will split conjugate eigenvalues. Request one cluster more or less. 
```
Any ideas how to automatically mitigate this? I noticced that using the eigen gap is not always very helpful to describe the variation in the system because the lower vectors may still contribute quite a bit if there are 1 or 2 isolated vectors with high eigenvalues, this is why I wanted to force it to use a fixed number of vectors.

**Comment by Marius1311** [comment]
Hi @davidsebfischer this is really a conceptual question. GPCCA needs real, invariant subspaces of the transition matrix P to compute the metastable decomposition. Let the real Schur decomposition be given by

P = Q R Q^T,

where Q is orthogonal and R is quasi upper-triangular, which means it has 1x1 and 2x2 blocks on the diagonal and is upper-triangular otherwise. 1x1 blocks directly correspond to real eigenvalues. 2x2 blocks are a real representation of a pair of complex conjugate eigenvalues. Note that for a real matrix P, complex eigenvalues by definition always appear in pairs of complex conjugates. Here comes the important bit: columns in Q corresponding to 1x1 blocks on the diagonal in R span invariant subspaces of P. However, columns corresponding to 2x2 blocks only span invariant subspaces **in pairs**, ie we need to consider both columns jointly. 

This means some cluster numbers are simply not possible by definition. Which ones these are depends on the transition matrix. The current behaviour is just to raise an exception in case the user requests such a cluster number. We can brainstorm alternative behaviours here.

**Comment by davidsebfischer** [comment]
thanks for the explanation @Marius1311! 
- Do I understand this correctly, that at the point at which you raise the error above, you already have a decomposition and check that the the first 3 (for example) vectors are not allowed because 3&4 have conjugate eigenvalues?
- If this is the case, you could just raise a warning and return a list of the allowed numbers of clusters, summarizing the information on the real and conjugate eigenvalues. A user or wrapping script could then use this list to rerun the function with the modified number of clusters. alternatively, to not recompute, you could allow a boolean setting that increments n_states until an allowed number is reached and just print this as a warning if this is done?

**Comment by Marius1311** [comment]
Hi David,
re 1: you're right, at the point where we raise the exception, we already have the decomposition
re 2: we could give out such a list already when running `.compute_schur()`, that way, the user would know before running the `.compute_metastable_states` function which values are allowed. In addition, we could increment the number of states by one in `.compute_metastable_states` in case the state number specified by the user is not valid.

@michalk8 , what's your opinion here?

**Comment by michalk8** [comment]
@Marius1311 agreed for both 1 and 2. For 2. we can also not allow available values (raise an error) and force user to manually to recompute the Schur decomposition instead of incrementing it by one (can again fail, although it's more convenient).

**Comment by Marius1311** [comment]
for the schur decomposition, I would always just increase it by one in case it fails. It can't fail again, because if the last eigenvalue was part of a pair of complex conjugates then the next one is going to complete this pair and we're fine.

**Comment by michalk8** [comment]
closed via #276

**Comment by Marius1311** [comment]
Fantastic! @davidsebfischer , you now get a list of admissible values when you call `estimator.compute_schur()`
